src/symmetric_crypto/aes_256_gcm_pure/mod.rs

```diff
@@ -3,73 +3,102 @@
 use aes_gcm::{
-    aead::{Aead, NewAead, Payload},
-    AeadInPlace, Aes256Gcm,
+    aead::{Aead, Payload},
+    aes::cipher::generic_array::GenericArray,
+    AeadInPlace, Aes256Gcm, KeyInit,
 };
-use generic_array::{typenum, GenericArray};
-use std::fmt::Display;
+use rand_core::{CryptoRng, RngCore};
 
-pub mod dem;
+use super::{key::Key, nonce::Nonce};
 
 /// Use a 256-bit AES key
-pub type KeyLength = typenum::U32;
+const KEY_LENGTH: usize = 32;
 
 /// Use a 96-bit nonce
-pub const NONCE_LENGTH: usize = 12;
+const NONCE_LENGTH: usize = 12;
 
 /// Use a 128-bit MAC tag
-pub const MAC_LENGTH: usize = 16;
+const MAC_LENGTH: usize = 16;
 
-pub type Key = crate::symmetric_crypto::key::Key<KeyLength>;
-
-pub type Nonce = crate::symmetric_crypto::nonce::Nonce<NONCE_LENGTH>;
+/// A 96-bit nonce restricts the plaintext size to 4096 bytes
+const MAX_PLAINTEXT_LENGTH: usize = 4096;
 
 /// Structure implementing `SymmetricCrypto` and the `DEM` interfaces based on
 /// AES 256 GCM.
+#[derive(Debug, PartialEq, Eq)]
 pub struct Aes256GcmCrypto;
 
-impl Display for Aes256GcmCrypto {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", Self::description())
-    }
-}
-
-impl SymmetricCrypto for Aes256GcmCrypto {
-    type Key = Key;
-    type Nonce = Nonce;
+impl Dem<KEY_LENGTH> for Aes256GcmCrypto {
+    const ENCRYPTION_OVERHEAD: usize = Self::Nonce::LENGTH + Self::MAC_LENGTH;
 
     const MAC_LENGTH: usize = MAC_LENGTH;
 
-    fn description() -> String {
-        format!(
-            "AES 256 GCM pure Rust (key bits: {}, nonce bits: {}, tag bits: {})",
-            KeyLength::to_usize() * 8,
-            NONCE_LENGTH * 8,
-            MAC_LENGTH * 8
-        )
-    }
+    type Key = Key<KEY_LENGTH>;
 
-    fn encrypt(
-        key: &Self::Key,
-        bytes: &[u8],
-        nonce: &Self::Nonce,
+    type Nonce = Nonce<NONCE_LENGTH>;
+
+    fn encrypt<R: RngCore + CryptoRng>(
+        rng: &mut R,
+        secret_key: &Self::Key,
+        plaintext: &[u8],
         additional_data: Option<&[u8]>,
     ) -> Result<Vec<u8>, CryptoCoreError> {
-        encrypt_combined(key, bytes, nonce, additional_data)
+        if plaintext.len() > MAX_PLAINTEXT_LENGTH {
+            return Err(CryptoCoreError::InvalidSize(format!(
+                "Plaintext is too large ({} bytes), max size: {} ",
+                plaintext.len(),
+                MAX_PLAINTEXT_LENGTH
+            )));
+        }
+        let nonce = Self::Nonce::new(rng);
+        // allocate correct byte number
+        let mut res: Vec<u8> = Vec::with_capacity(plaintext.len() + Self::ENCRYPTION_OVERHEAD);
+        res.extend_from_slice(nonce.as_bytes());
+        res.append(
+            &mut encrypt_combined(
+                secret_key.as_bytes(),
+                plaintext,
+                nonce.as_bytes(),
+                additional_data,
+            )
+            .map_err(|err| CryptoCoreError::EncryptionError(err.to_string()))?,
```

Comment:
Why mapping CryptoCoreError to CryptoCoreError?

Reply:
Indeed, I didn't see it